src/lwc/geFormWidgetTokenizeCard/geFormWidgetTokenizeCard.js

```diff
@@ -1,18 +1,114 @@
+        this.registerPostMessageListener();
+    }
+
+    /*******************************************************************************
+    * @description Builds the visualforce origin url that we need in order to
+    * make sure we're only listening for messages from the correct source in the
+    * registerPostMessageListener method.
+    */
+    buildVisualforceOriginUrl(domain) {
+        let url = `https://${domain}--c.visualforce.com`
+        if (TemplateBuilderService.namespaceWrapper) {
+            const currentNamespace = TemplateBuilderService.namespaceWrapper.currentNamespace;
+
+            if (currentNamespace) {
+                url = url.replace('--c', `--${currentNamespace}`);
+            }
+        }
+
+        return url;
+    }
+
+    /*******************************************************************************
+    * @description Method listens for a message from the visualforce iframe.
+    * Rejects any messages from an unknown origin.
+    */
+    registerPostMessageListener() {
+        let component = this;
+
+        window.onmessage = async function (event) {
+            if (event && event.origin !== component.visualforceOrigin) {
+                // Reject any messages from an unexpected origin
+                return;
+            } else {
+                const message = JSON.parse(event.data);
+                component.handleMessage(message);
+            }
+        }
+    }
+
+    /*******************************************************************************
+    * @description Method handles messages received from iframed visualforce page.
+    *
+    * @param {object} message: Message received from iframe
+    */
+    async handleMessage(message) {
+        if (message.error) {
+            // Error with tokenization
+            let error = JSON.stringify(message.error);
+            console.log(error);
+            alert(error);
+        } else if (message.token) {
+            // TODO: Start - Remove later
+            // Make purchase call... for dev only
+            try {
+                let purchaseCallResponse = await makePurchaseCall({ token: message.token });
+                this.purchaseResult = JSON.parse(purchaseCallResponse);
+                console.log(this.purchaseResult);
+                alert(this.purchaseResult)
+            } catch (error) {
+                handleError(error);
+            }
+            // TODO: End - Remove later
+
+            // TODO: Save token locally in widget until form requests it
+        } else if (message.isLoaded) {
+            this.isLoading = false;
+        }
+    }
+
+    /*******************************************************************************
+    * @description Method sends a message to the visualforce page iframe requesting
+    * a token. Response for this request is found and handled in
+    * registerPostMessageListener.
+    */
+    requestToken() {
+        const iframe = this.template.querySelector('iframe');
```

Comment:
Should we/can we add a "data-" attribute and use that as a selector?  (although the likelihood of having another iframe in the template is low.... so maybe not necessary)

Reply:
Adding that in. This should be the only iframe element that's visible to the component, but I think better practice to use a specific identifier regardless 🙏